src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1413 @@
+            private readonly TypeInferenceContext _context;
+            internal ContextPermissionScope(TypeInferenceContext context, TypeInferenceRuntimePermissions scopePermissions)
+            {
+                _context = context;
+                _permissions = context.RuntimePermissions;
+                _context.RuntimePermissions = scopePermissions;
+            }
+
+            void IDisposable.Dispose()
+            {
+                _context.RuntimePermissions = _permissions;   
+            }
+        }
+    }
+
+    internal class TypeInferenceContext
+    {
+        public static readonly PSTypeName[] EmptyPSTypeNameArray = Utils.EmptyArray<PSTypeName>();
+        private readonly PowerShell _powerShell;
+        internal PowerShellExecutionHelper Helper { get; }
+
+        public TypeInferenceContext() : this(PowerShell.Create(RunspaceMode.CurrentRunspace))
+        {
+        }
+
+        public TypeInferenceContext(PowerShell powerShell)
+        {
+            _powerShell = powerShell;
+            Helper = new PowerShellExecutionHelper(powerShell);
+
+        }
+
+        public TypeDefinitionAst CurrentTypeDefinitionAst { get; set; }
+        public TypeInferenceRuntimePermissions RuntimePermissions { get; set; }
+        internal ExecutionContext ExecutionContext => _powerShell.Runspace.ExecutionContext;
+
+        internal IEnumerable<PSTypeName> InferType(Ast ast, TypeInferenceVisitor visitor)
+        {
+            var res = ast.Accept(visitor);
+            if (res == null)
+            {
+                return EmptyPSTypeNameArray;
+            }
+            return (IEnumerable<PSTypeName>)res;
+        }
+
+        public bool TryGetRepresentativeTypeNameFromExpressionSafeEval(ExpressionAst expression, out PSTypeName typeName)
+        {
+            typeName = null;
+            if (!RuntimePermissions.HasFlag(TypeInferenceRuntimePermissions.AllowSafeEval)) return false;
+            object value;
+            return expression != null &&
+                   SafeExprEvaluator.TrySafeEval(expression, ExecutionContext, out value) &&
+                   TryGetRepresentativeTypeNameFromValue(value, out typeName);
+        }
+
+        internal static bool TryGetRepresentativeTypeNameFromValue(object value, out PSTypeName type)
+        {
+            type = null;
+            if (value != null)
+            {
+                var list = value as IList;
+                if (list != null && list.Count > 0)
+                {
+                    value = list[0];
+                }
+                value = PSObject.Base(value);
+                if (value != null)
+                {
+                    type = new PSTypeName(value.GetType());
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public IList<object> GetMembersByInferredType(PSTypeName typename, bool @static, Func<object, bool> filter)
```

Comment:
Please, avoid using `@static`, `@class`, etc for variable names. Replace by `isStatic`.